openmc/source.py

```diff
@@ -330,12 +330,47 @@ def from_xml_element(cls, elem: ET.Element) -> 'openmc.Source':
         return source
 
 
-class ParticleType(Enum):
+class ParticleType(IntEnum):
+    """
+    IntEnum class representing a particle type. Type
+    values mirror those found in the C++ class.
+    """
     NEUTRON = 0
     PHOTON = 1
     ELECTRON = 2
     POSITRON = 3
 
+    @classmethod
+    def from_string(cls, value: str):
+        """
+        Constructs a ParticleType instance from a string.
+
+        Parameters
+        ----------
+        value : str
+            The string representation of the particle type.
+
+        Returns
+        -------
+        The corresponding ParticleType instance.
+        """
+        try:
+            return cls[value.upper()]
+        except KeyError:
+            raise ValueError(f"Invalid string for creation of {cls.__name__}: {value}")
+
+    def __repr__(self) -> str:
+        """
+        Returns a string representation of the ParticleType instance.
+
+        Returns:
+            str: The lowercase name of the ParticleType instance.
+        """
+        return self.name.lower()
+
+    def __str__(self) -> str:
```

Comment:
It turns out that `__str__` won't default to `__repr__` for the `IntEnum` class. This is being updated, but won't be effective until Python 3.11 ([PEP663](https://peps.python.org/pep-0663/)). I can make a note here to drop in the future if you like.

Reply:
I also had to override the default `__format__` method in this class for our Python 3.7 tests. Once we move past 3.7 we can remove it.